File: done/AsteroidCollision.java
package done;

import java.util.Stack;

public class AsteroidCollision {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for (int asteroid : asteroids) {
            while (!stack.isEmpty() && asteroid < 0 && stack.lastElement() > 0) {
                int dif = asteroid + stack.lastElement();
                if (dif > 0) {
                    asteroid = 0;
                }
                else if (dif < 0) {
                    stack.pop();
                }
                else {
                    stack.pop();
                    asteroid = 0;
                }
            }
            if (asteroid != 0) {
                stack.push(asteroid);
            }
        }

        return stack.stream().mapToInt(Integer::intValue).toArray();
    }

    public int[] asteroidCollision2(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for (int asteroid : asteroids) {
            while (!stack.isEmpty() && stack.lastElement() > 0 && stack.lastElement() < -asteroid) {
                stack.pop();
            }

            if (stack.isEmpty() || asteroid > 0 || stack.lastElement() < 0) {
                stack.push(asteroid);
            }
            else if (asteroid < 0 && stack.lastElement() == -asteroid) {
                stack.pop();
            }
        }

        return stack.stream().mapToInt(Integer::intValue).toArray();
    }

}

File: done/AsteroidCollision.kt
import java.util.*

fun main() {
    val arr = asteroidCollision(intArrayOf(-5,5,10,-5))
//    val arr = asteroidCollision(intArrayOf(5,10,-5))
//    val arr = asteroidCollision(intArrayOf(8,-8))
//    val arr = asteroidCollision(intArrayOf(10,2,-5))
    for (i in arr)
        println(i)
}

fun asteroidCollision(asteroids: IntArray): IntArray {
    val stack = Stack<Int>()

    for (i in asteroids) {
        while (stack.isNotEmpty() && stack.lastElement() > 0 && stack.lastElement() < -i)
            stack.pop()

        if (stack.isEmpty() || i > 0 || stack.lastElement() < 0)
            stack.push(i)
        else if (i < 0 && stack.lastElement() == -i)
            stack.pop()
    }

    return stack.toIntArray()
}

